src/main/java/org/opensearch/ad/ratelimit/RateLimitedQueue.java

```diff
@@ -0,0 +1,474 @@
+                        return Optional.of(requests);
+                    }
+                }
+            }
+            // if we haven't find a non-empty queue , return empty.
+            return Optional.empty();
+        } finally {
+            // it is fine we may have race conditions. We are not trying to
+            // be precise. The objective is to select each segment with equal probability.
+            lastSelectedSegmentId = startId;
+        }
+    }
+
+    private void clearExpiredRequests(BlockingQueue<RequestType> requests) {
+        // In terms of request duration, HCAD throws a request out if it
+        // is older than the detector frequency. This duration limit frees
+        // up HCAD to work on newer requests in the subsequent detection
+        // interval instead of piling up requests that no longer matter.
+        // For example, loading model checkpoints for cache misses requires
+        // a queue configured in front of it. A request contains the checkpoint
+        // document Id and the expiry time, and the queue can hold a considerable
+        // volume of such requests since the size of the request is small.
+        // The expiry time is the start timestamp of the next detector run.
+        // Enforcing the expiry time places an upper bound on each request’s
+        // lifetime.
+        RequestType head = requests.peek();
+        while (head != null && head.getExpirationEpochMs() < clock.millis()) {
+            requests.poll();
+            head = requests.peek();
+        }
+    }
+
+    protected void putOnly(RequestType request) {
+        try {
+            Segment requestQueue = requestSegments
+                .computeIfAbsent(
+                    SegmentPriority.MEDIUM == request.getPriority() ? request.getDetectorId() : request.getPriority().name(),
+                    k -> new Segment()
+                );
+
+            requestQueue.setLastAccessTime(clock.instant());
+            requestQueue.put(request);
+        } catch (Exception e) {
+            LOG.error(new ParameterizedMessage("Failed to add requests to [{}]", this.queueName), e);
+        }
+    }
+
+    private void maintainForThreadPool() {
+        for (final ThreadPoolStats.Stats stats : threadPool.stats()) {
+            String name = stats.getName();
+            // cold entity queue mostly use these 3 threadpools
+            if (ThreadPool.Names.SEARCH.equals(name) || ThreadPool.Names.GET.equals(name) || ThreadPool.Names.WRITE.equals(name)) {
+                if (stats.getQueue() > (int) (maxQueuedTaskRatio * threadPool.info(name).getQueueSize().singles())) {
+                    setCoolDownStart();
+                    break;
+                }
+            }
+        }
+    }
+
+    private void prune(Map<String, Segment> segments) {
+        for (Map.Entry<String, Segment> segmentEntry : segments.entrySet()) {
+            if (segmentEntry.getKey().equals(SegmentPriority.HIGH.name())) {
+                return;
```

Comment:
Why does this return immediately before checking other priorities?

Reply:
good catch. changed to continue.